### PR TITLE
[FEAT] add aio support

### PR DESCRIPTION
**Pull request by mordajon**
*Async queries is done natively using aiohttp behind the scenes*
*Async authentication is done using ThreadExecutor due to library dependencies*

1. Added aio API and transform all derivation to be none blocking
2. Previous sync API left the same way while extracting common logic into reusable functions
3. Added dependency in aiohttp, asgiref, aiofile when installing optional async
   - aiohttp (for native async http calls)
   - asgiref (msrestazure, adal, azure-cli - are blocking libraries, so we need to wrap all of them in executor)
   - aiofile (azure cli authentication reads from file, doing it async now)
4. Added static typings
5. Added async installation to travis
6. Async support was added according to [azure guidelines](https://azure.github.io/azure-sdk/python_design.html#async-support)
7. Modified ReadMe to show examples of usage
8. Added helper decorators to avoid documentation duplication
9. Modified .travis.yml matrix

### CODE REVIEW

**Comment by mordajon** [comment]
Usage example:
image

Output Example:
image

**Comment by danield137** [comment]
https://github.com/Azure/azure-kusto-python/issues/153

File: azure-kusto-data/azure/kusto/data/request.py [hunk 1]
```diff
@@ -506,6 +512,7 @@ def __init__(self, kcsb: Union[KustoConnectionStringBuilder, str]):
         kusto_cluster = kcsb.data_source
 
         # Create a session object for connection pooling
+        # self._session is only used using sync access
```
Comment (danield137) [inline azure-kusto-data/azure/kusto/data/request.py]: why? can't use session in async flow?

File: azure-kusto-data/azure/kusto/data/request.py [hunk 2]
```diff
@@ -516,7 +523,20 @@ def __init__(self, kcsb: Union[KustoConnectionStringBuilder, str]):
         self._auth_provider = _AadHelper(kcsb) if kcsb.aad_federated_security else None
         self._request_headers = {"Accept": "application/json", "Accept-Encoding": "gzip,deflate", "x-ms-client-version": "Kusto.Python.Client:" + VERSION}
 
+    @staticmethod
+    def _build_runtime_error(func_name: str) -> RuntimeError:
```
Comment (danield137) [inline azure-kusto-data/azure/kusto/data/request.py]: I dislike build functions. these are ctors, and should be implemented as such. Also, since this is an expected error, we should probably raise it with a `Kusto` prefix like `KustoAsyncRuntimeError`

File: azure-kusto-data/azure/kusto/data/request.py [hunk 3]
```diff
@@ -516,7 +523,20 @@ def __init__(self, kcsb: Union[KustoConnectionStringBuilder, str]):
...
+        return RuntimeError(f"Kusto driver has detected a running event loop, consider using {func_name}_async instead of {func_name}")
+
     def execute(self, database: str, query: str, properties: ClientRequestProperties = None) -> KustoResponseDataSet:
+        """
+        Sync version of self.execute_async
+        """
+        try:
+            return async_to_sync(self.execute_async)(database, query, properties)
+        except RuntimeError as e:
+            raise self._build_runtime_error("execute") from e
```
Comment (danield137) [inline azure-kusto-data/azure/kusto/data/request.py]: this means that running sync in a non-async env would break?

File: azure-kusto-data/azure/kusto/data/_version.py [hunk 4]
```diff
@@ -1,2 +1,2 @@
 """Version file."""
-VERSION = "0.0.43"
+VERSION = "0.0.44"
```
Comment (danield137) [inline azure-kusto-data/azure/kusto/data/_version.py]: this is only done when releasing a version.. no need

File: azure-kusto-data/azure/kusto/data/asyncio/utils.py [hunk 5]
```diff
@@ -0,0 +1,9 @@
+import aiohttp
+
+
+async def run_request(endpoint: str, **kwargs) -> dict:
```
Comment (danield137) [inline azure-kusto-data/azure/kusto/data/asyncio/utils.py]: I dislike utils classes. I prefer you put this function close to it's usage (kusto doesn't provide utils)

File: azure-kusto-data/setup.py [hunk 6]
```diff
@@ -41,6 +41,6 @@
     namespace_packages=["azure"],
     keywords="kusto wrapper client library",
     packages=find_packages(exclude=["azure", "tests"]),
-    install_requires=["adal>=1.0.0", "python-dateutil>=2.8.0", "requests>=2.13.0", "msrestazure>=0.4.14"],
+    install_requires=["adal>=1.0.0", "python-dateutil>=2.8.0", "requests>=2.13.0", "msrestazure>=0.4.14", "aiohttp==3.6.2", "asgiref==3.2.3", "aiofile==1.5.2"],
```
Comment (danield137) [inline azure-kusto-data/setup.py]: these should be optionals. (like the pandas thing). people using sync flow don't need them.

File: azure-kusto-data/azure/kusto/data/request.py (hunk 3 above)
Comment (mordajon) [reply azure-kusto-data/azure/kusto/data/request.py]: Nope... it means that when trying to run sync flow from async env it would break...

**Comment by mordajon** [comment]
Tests will be added upon API approval by @danield137

File: azure-kusto-data/azure/kusto/data/request.py (hunk 1 above)
Comment (mordajon) [reply azure-kusto-data/azure/kusto/data/request.py]: Async operations are wrapped by aiohttp.ClientSession which is way way faster:
image
https://julien.danjou.info/python-and-fast-http-clients/

File: azure-kusto-data/azure/kusto/data/request.py (hunk 2 above)
Comment (mordajon) [reply azure-kusto-data/azure/kusto/data/request.py]: 👍

File: azure-kusto-data/azure/kusto/data/_version.py (hunk 4 above)
Comment (mordajon) [reply azure-kusto-data/azure/kusto/data/_version.py]: 👍

File: azure-kusto-data/setup.py (hunk 6 above)
Comment (mordajon) [reply azure-kusto-data/setup.py]: Ammm... I am not sure I am following you with this? why are they optional?

File: azure-kusto-data/azure/kusto/data/asyncio/utils.py (hunk 5 above)
Comment (mordajon) [reply azure-kusto-data/azure/kusto/data/asyncio/utils.py]: 👍

File: azure-kusto-data/setup.py (hunk 6 above)
Comment (danield137) [reply azure-kusto-data/setup.py]: In theory, installing these packages takes time and space.
TBH, I'm still not sure I want async flow to be the default right now.

I want people to **opt-in** into this flow.
This means that people that want to use the regular flow, still can without extra cost.

File: .travis.yml [hunk 7]
```diff
@@ -17,7 +17,7 @@ install:
   - pip install -U pip
   - pip install -r dev_requirements.txt
   - pip install codecov coverage
-  - pip install ./azure-kusto-data ./azure-kusto-ingest
+  - pip install ./azure-kusto-data[async] ./azure-kusto-ingest
```
Comment (danield137) [inline .travis.yml]: Please revert.
Testing the sync flow without dependencies is more important.
That being said, we need to come up with a way to test both scenarios independently.

File: azure-kusto-data/azure/kusto/data/request.py [hunk 8]
```diff
@@ -530,6 +546,18 @@ def execute(self, database: str, query: str, properties: ClientRequestProperties
             return self.execute_mgmt(database, query, properties)
         return self.execute_query(database, query, properties)
 
+    async def execute_async(self, database: str, query: str, properties: ClientRequestProperties = None) -> KustoResponseDataSet:
```
Comment (danield137) [inline azure-kusto-data/azure/kusto/data/request.py]: how does python handle the async keyword if no asynio is installed?

File: azure-kusto-data/azure/kusto/data/security.py [hunk 9]
```diff
@@ -116,32 +150,61 @@ def __init__(self, kcsb):
         aad_authority_uri = os.environ.get("AadAuthorityUri", CLOUD_LOGIN_URL)
         self.authority_uri = aad_authority_uri + authority if aad_authority_uri.endswith("/") else aad_authority_uri + "/" + authority
 
-    def acquire_authorization_header(self):
+    def _build_kusto_authentication_error(self, error: KustoAuthenticationError) -> KustoAuthenticationError:
```
Comment (danield137) [inline azure-kusto-data/azure/kusto/data/security.py]: should be a `KustoAuthenticationError` ctor / factory method

File: azure-kusto-data/azure/kusto/data/security.py [hunk 10]
```diff
@@ -198,15 +253,81 @@ def _acquire_authorization_header(self) -> str:
...
+        if self.authentication_method == AuthenticationMethod.az_cli_profile:
+            stored_token = await _get_azure_cli_auth_token_async()
+            self._set_from_stored_token(stored_token)
+            refresh_token = stored_token[TokenResponseFields.REFRESH_TOKEN]
+
+        if self.auth_context is None:
+            self.auth_context = AuthenticationContext(self.authority_uri)
+
+        if refresh_token is not None:
+            token = await sync_to_async(self.auth_context.acquire_token_with_refresh_token)(refresh_token, self.client_id, self.kusto_uri)
```
Comment (danield137) [inline azure-kusto-data/azure/kusto/data/security.py]: I'm not sure this is safe. Adal's AuthContext was not built with multi-threading in mind.

File: azure-kusto-data/tests/async_unittest.py [hunk 11]
```diff
@@ -0,0 +1,11 @@
+import unittest
```
Comment (danield137) [inline azure-kusto-data/tests/async_unittest.py]: please use pytest (unitest code is legacy code)

File: azure-kusto-data/tests/test_kusto_client.py [hunk 12]
```diff
@@ -341,43 +327,186 @@ def test_admin_then_query(self, mock_post):
...
         assert row[3] is None
-
         assert isinstance(row[4], str)
         assert row[4] == '{"rowId":2,"arr":[0,2]}'
-
         assert isinstance(row[5], dict)
         assert row[5] == {"rowId": 2, "arr": [0, 2]}
 
+
+class KustoClientTestsSync(unittest.TestCase, KustoClientMixin):
```
Comment (danield137) [inline azure-kusto-data/tests/test_kusto_client.py]: Please revert this. I'm not a fan of splitting assertion code from tests.
There are other ways of reducing code.
(also, this is unrelated to the PR)

File: azure-kusto-data/tests/test_kusto_client.py [hunk 13]
```diff
@@ -341,43 +327,186 @@ def test_admin_then_query(self, mock_post):
...
+        query = """print dynamic(123), dynamic("123"), dynamic("test bad json"),"""
+        """ dynamic(null), dynamic('{"rowId":2,"arr":[0,2]}'), dynamic({"rowId":2,"arr":[0,2]})"""
+        row = client.execute_query("PythonTest", query).primary_results[0].rows[0]
+        self._assert_dynamic_response(row)
+
     @patch("requests.Session.post", side_effect=mocked_requests_post)
     def test_empty_result(self, mock_post):
         """Tests dynamic responses."""
-        client = KustoClient("https://somecluster.kusto.windows.net")
+        client = self._create_kusto_client()
```
Comment (danield137) [inline azure-kusto-data/tests/test_kusto_client.py]: revert. I believe in explicit creation over implicit, this code hides the actual value for no real reason.

File: azure-kusto-data/tests/test_kusto_client.py [hunk 14]
```diff
@@ -341,43 +327,186 @@ def test_admin_then_query(self, mock_post):
...
+        self._mock_query(mocked)
+        client = self._create_kusto_client()
+        response = self.loop.run_until_complete(client.execute_query_async("PythonTest", "Deft"))
+        self._assert_sanity_query_response(response)
+
+    @aioresponses()
+    def test_sanity_control_command_async(self, mocked):
+        """Async version of self.test_sanity_control_command"""
+        self._mock_mgmt(mocked)
+        client = self._create_kusto_client()
```
Comment (danield137) [inline azure-kusto-data/tests/test_kusto_client.py]: pytest (and me) are against test classes. you should refrain from using `self` in tests

File: dev_requirements.txt [hunk 15]
```diff
@@ -2,4 +2,6 @@ pytest>=3.2.0
 mock>=2.0.0
 responses>=0.9.0
 pandas>=0.24.0
-black;python_version == '3.6'
\ No newline at end of file
+black;python_version == '3.6'
+aioresponses==0.6.2
```
Comment (danield137) [inline dev_requirements.txt]: like the above, and previous discussion, I do not intend for asyn flow to be the main flow. as such, we cannot declare it as an equal rights citizen everwhere. it should always be Opt-In.

**Comment by danield137** [comment]
After the last changes, I'm starting to think a better way would be to just have a separate class entirely.

`KustoClient` vs `KustoAsyncClient`

This will create the cleanest separation possible.

Also, one thing I did not consider is that for async flow to work well, everything needs to be async.
Since that is not the case (adal) I am not sure what are the implication of faking it.

**Comment by mordajon** [comment]
> After the last changes, I'm starting to think a better way would be to just have a separate class entirely.
> 
> `KustoClient` vs `KustoAsyncClient`
> 
> This will create the cleanest separation possible.
> 
> Also, one thing I did not consider is that for async flow to work well, everything needs to be async.
> Since that is not the case (adal) I am not sure what are the implication of faking it.

A lot of libraries out there wrapping sync actions with thread executor. 
We only wrap the authentication which happens once per... 

Also I am working with ADAL team as well to create an async flow there as well.
We can use it in the next version 😊

File: azure-kusto-data/azure/kusto/data/request.py (hunk 8 above)
Comment (mordajon) [reply azure-kusto-data/azure/kusto/data/request.py]: Asyncio comes integrated with python 3.5+

File: azure-kusto-data/tests/async_unittest.py (hunk 11 above)
Comment (mordajon) [reply azure-kusto-data/tests/async_unittest.py]: I agree, pytest is way better... But I wanted to keep the same writing style

File: azure-kusto-data/tests/test_kusto_client.py (hunk 12 above)
Comment (mordajon) [reply azure-kusto-data/tests/test_kusto_client.py]: It is related because async flow uses the same asserts

File: dev_requirements.txt (hunk 15 above)
Comment (mordajon) [reply dev_requirements.txt]: What is your suggestion in regards of dev_requirements?

Comment (mordajon) [reply dev_requirements.txt]: I believe in tests we should check all of the flows, don't you think?

File: .travis.yml (hunk 7 above)
Comment (mordajon) [reply .travis.yml]: Amm well I think that tests should cover everything... to prevent code changes that will break async or panda for that matter

File: azure-kusto-data/azure/kusto/data/security.py (hunk 10 above)
Comment (mordajon) [reply azure-kusto-data/azure/kusto/data/security.py]: From asgi docs:
"If the outermost program is async (i.e. SyncToAsync is outermost), then
    this will be a dedicated single sub-thread that all sync code runs in,
    one after the other"

File: azure-kusto-data/tests/test_kusto_client.py (hunk 14 above)
Comment (mordajon) [reply azure-kusto-data/tests/test_kusto_client.py]: I agree that pytest is the better approach...

But the way I see it, we either change test_kusto_client to be written in pytest completely or keep it the way it is today.

My thought was that you won't want to see a lot of changes in the tests (to make sure everything runs exactly like before)

File: dev_requirements.txt (hunk 15 above)
Comment (mordajon) [reply dev_requirements.txt]: do mind... that pandas is in that file as well

Comment (danield137) [reply dev_requirements.txt]: there is a large difference between the two, but I agree that pandas should also be out of these tests.

**Comment by mordajon** [comment]
Creating another commit to follow Azure guidelines:
https://azure.github.io/azure-sdk/python_design.html#async-support

File: azure-kusto-data/azure/kusto/data/_models.py [hunk 16]
```diff
@@ -16,6 +18,18 @@
...
     HAS_PANDAS = False
 
+aio_installed = True
+try:
+    from aiohttp import ClientResponse, ClientSession
+except ImportError:
+    aio_installed = False
+
+
+def _get_dynamic_response_type():
```
Comment (danield137) [inline azure-kusto-data/azure/kusto/data/_models.py]: where do you use this?

plus, not sure why you need this code here, everything aio is under subfolder

File: azure-kusto-data/azure/kusto/data/aio/request.py [hunk 17]
```diff
@@ -0,0 +1,75 @@
...
+
+try:
+    from aiohttp import ClientResponse, ClientSession
+except ImportError:
+    raise KustoAioSyntaxError()
+
+ResponseTuple = namedtuple("ResponseTuple", ["response", "response_json"])
+
+
+@documented_by(KustoClientSync)
```
Comment (danield137) [inline azure-kusto-data/azure/kusto/data/aio/request.py]: please don't do this.
it blocks the ability to render this file to a readthedocs page. comments should be static

File: azure-kusto-data/azure/kusto/data/request.py [hunk 18]
```diff
@@ -1,631 +1,657 @@
...
-            return KustoResponseDataSetV1(response.json())
-
-        if payload:
-            raise KustoServiceError(
-                "An error occurred while trying to ingest: Status: {0.status_code}, Reason: {0.reason}, Text: {0.text}".format(response), response
-            )
-
-        raise KustoServiceError([response.json()], response)
+"""A module to make a Kusto request."""
+import io
```
Comment (danield137) [inline azure-kusto-data/azure/kusto/data/request.py]: did you change line endings or something? the entire file shouldn't be changed

File: azure-kusto-data/azure/kusto/data/request.py [hunk 19]
```diff
@@ -1,631 +1,657 @@
...
+                request_headers["x-ms-app"] = properties.application
+            if properties.user is not None:
+                request_headers["x-ms-user"] = properties.user
+        if properties:
+            timeout = properties.get_option(ClientRequestProperties.request_timeout_option_name, timeout)
+
+        return RequestParams(json_payload, request_headers, timeout)
+
+    @staticmethod
+    def _build_kusto_service_error(payload: io.IOBase, response: _get_dynamic_response_type(), response_json: List[dict]) -> KustoServiceError:
```
Comment (danield137) [inline azure-kusto-data/azure/kusto/data/request.py]: already commented about this: this should be a ctor for the kustoserviceerror

Comment (danield137) [inline azure-kusto-data/azure/kusto/data/request.py]: this `_get_dynamic_response_type` beats the purpose of annotations. annotations are here to help me read the code and make sure I don't make typing mistakes. if they are resolved in runtime it does not help at all.
please revert this.

if you want a dynamic type, just use `Any` and duck type it

File: azure-kusto-data/azure/kusto/data/request.py [hunk 20]
```diff
@@ -1,631 +1,657 @@
...
+    def get_option(self, name, default_value):
+        """Gets an option's value."""
+        return self._options.get(name, default_value)
+
+    def to_json(self):
+        """Safe serialization to a JSON string."""
+        return json.dumps({"Options": self._options, "Parameters": self._parameters}, default=str)
+
+
+RequestParams = namedtuple("RequestParams", ["json_payload", "request_headers", "timeout"])
```
Comment (danield137) [inline azure-kusto-data/azure/kusto/data/request.py]: underscore (mark as private)

File: azure-kusto-data/azure/kusto/data/security.py [hunk 21]
```diff
@@ -116,57 +79,100 @@ def __init__(self, kcsb):
...
+            self.client_id = stored_token[TokenResponseFields._CLIENT_ID]
+            self.username = stored_token[TokenResponseFields.USER_ID]
+            self.authority_uri = stored_token[TokenResponseFields._AUTHORITY]
+
+    @staticmethod
+    def _get_expiration_data_from_token(token):
+        return dateutil.parser.parse(token[TokenResponseFields.EXPIRES_ON])
+
+    @staticmethod
+    def _authentication_method_missing_exception() -> KustoClientError:
```
Comment (danield137) [inline azure-kusto-data/azure/kusto/data/security.py]: don't wrap error creation. all over the place.

File: azure-kusto-data/azure/kusto/data/security.py [hunk 22]
```diff
@@ -116,57 +79,100 @@ def __init__(self, kcsb):
...
+        return KustoClientError("Please choose authentication method from azure.kusto.data.security.AuthenticationMethod")
+
+    @staticmethod
+    def _is_expired(expiration_date):
```
Comment (danield137) [inline azure-kusto-data/azure/kusto/data/security.py]: don't hide the logic. nobody knows what expired means. this method should be removed

File: azure-kusto-data/tests/aio/test_kusto_client.py [hunk 23]
```diff
@@ -0,0 +1,127 @@
...
+    import asgiref
+
+    aio_installed = True
+except:
+    pass
+
+
+@pytest.mark.skipif(not aio_installed, reason="requires aio")
+@aio_documented_by(KustoClientTestsSync)
+class KustoClientTestsTests(TestCase, KustoClientTestsMixin):
```
Comment (danield137) [inline azure-kusto-data/tests/aio/test_kusto_client.py]: naming? why the double tests?

File: azure-kusto-data/azure/kusto/data/security.py (hunk 9 above)
Comment (mordajon) [reply azure-kusto-data/azure/kusto/data/security.py]: I think this is beyond of the scope of this task, and this PR is very big already... let's do it later

File: azure-kusto-data/azure/kusto/data/request.py (hunk 18 above)
Comment (mordajon) [reply azure-kusto-data/azure/kusto/data/request.py]: Not sure what do you mean by this

File: azure-kusto-data/azure/kusto/data/request.py (hunk 19 above)
Comment (mordajon) [reply azure-kusto-data/azure/kusto/data/request.py]: I prefered to duplicate this specific method, so the typing will be more accurate and informative

File: azure-kusto-data/azure/kusto/data/security.py (hunk 22 above)
Comment (mordajon) [reply azure-kusto-data/azure/kusto/data/security.py]: I prefer that the logic of expiration will be written once... and not twice

**Comment by mordajon** [comment]
ExecuteRequestParams should take authorization as param in addition to self._request_headers

File: .travis.yml [hunk 24]
```diff
@@ -1,23 +1,24 @@
...
-  fast_finish: true
+os:
+  - linux
+env:
+  - OPTIONAL=None
+  - OPTIONAL=aio
+  - OPTIONAL=pandas
+python:
+  - "3.5"
+  - "3.6"
```
Comment (danield137) [inline .travis.yml]: 🎉

File: azure-kusto-data/azure/kusto/data/security.py (hunk 9 above)
Comment (danield137) [reply azure-kusto-data/azure/kusto/data/security.py]: you can have a ctor that accepts the helper itself..

File: azure-kusto-data/azure/kusto/data/security.py (hunk 22 above)
Comment (danield137) [reply azure-kusto-data/azure/kusto/data/security.py]: if you want to, then you need to promote time to be a parameter, with a default value like:
`is_expired(expiration_date, expiry_time=1 * 60`

I'm sure you're familiar with python's zen (explicit is better than implicit)

**Comment by roeap** [comment]
Really looking forward for this feature to become available. Is there a plan to merge this PR soon?

**Comment by woanware** [comment]
Have been testing the PR and it works in my scenario's, would be great to get it merged!

**Comment by woanware** [comment]
Since using the PR code (under Windows 10, Python 3.7.6 x64) I have noticed regular asyncio errors, when running multiple async queries, for example:

```---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
C:\Python37\lib\runpy.py in _run_module_as_main(mod_name, alter_argv)
    191         sys.argv[0] = mod_spec.origin
    192     return _run_code(code, main_globals, None,
--> 193                      "__main__", mod_spec)
    194 
    195 def run_module(mod_name, init_globals=None,

C:\Python37\lib\runpy.py in _run_code(code, run_globals, init_globals, mod_name, mod_spec, pkg_name, script_name)
     83                        __package__ = pkg_name,
     84                        __spec__ = mod_spec)
---> 85     exec(code, run_globals)
     86     return run_globals
     87 

D:\Dev\venv\lib\site-packages\ipykernel_launcher.py in <module>
     14 
     15     from ipykernel import kernelapp as app
---> 16     app.launch_new_instance()

D:\Dev\venv\lib\site-packages\traitlets\config\application.py in launch_instance(cls, argv, **kwargs)
    662         app = cls.instance(**kwargs)
    663         app.initialize(argv)
--> 664         app.start()
    665 
    666 #-----------------------------------------------------------------------------

D:\Dev\venv\lib\site-packages\ipykernel\kernelapp.py in start(self)
    581         self.io_loop = ioloop.IOLoop.current()
    582         try:
--> 583             self.io_loop.start()
    584         except KeyboardInterrupt:
    585             pass

D:\Dev\venv\lib\site-packages\tornado\platform\asyncio.py in start(self)
    130             self._setup_logging()
    131             asyncio.set_event_loop(self.asyncio_loop)
--> 132             self.asyncio_loop.run_forever()
    133         finally:
    134             asyncio.set_event_loop(old_loop)

C:\Python37\lib\asyncio\base_events.py in run_forever(self)
    536             events._set_running_loop(self)
    537             while True:
--> 538                 self._run_once()
    539                 if self._stopping:
    540                     break

C:\Python37\lib\asyncio\base_events.py in _run_once(self)
   1744                            timeout * 1e3, dt * 1e3)
   1745         else:
-> 1746             event_list = self._selector.select(timeout)
   1747         self._process_events(event_list)
   1748 

C:\Python37\lib\selectors.py in select(self, timeout)
    321         ready = []
    322         try:
--> 323             r, w, _ = self._select(self._readers, self._writers, [], timeout)
    324         except InterruptedError:
    325             return ready

C:\Python37\lib\selectors.py in _select(self, r, w, _, timeout)
    312     if sys.platform == 'win32':
    313         def _select(self, r, w, _, timeout=None):
--> 314             r, w, x = select.select(r, w, w, timeout)
    315             return r, w + x, []
    316     else:

ValueError: too many file descriptors in select()
```

**Comment by woanware** [comment]
To fix the "too many file selectors" error I had to change the following line, since it kept the HTTP socket open:

```
self._request_headers = {"Accept": "application/json", "Accept-Encoding": "gzip,deflate", "x-ms-client-version": "Kusto.Python.Client:" + VERSION}
```

To:

```
self._request_headers = {'Connection':'close', "Accept": "application/json", "Accept-Encoding": "gzip,deflate", "x-ms-client-version": "Kusto.Python.Client:" + VERSION}
```

**Comment by woanware** [comment]
Bump?

**Comment by vladikbr** [comment]
Already committed
